massa-graph-2-worker/src/controller.rs

```diff
@@ -0,0 +1,136 @@
+use massa_graph::{
+    error::{GraphError, GraphResult},
+    export_active_block::ExportActiveBlock,
+    BootstrapableGraph,
+};
+use massa_graph_2_exports::{
+    block_graph_export::BlockGraphExport, block_status::BlockStatus, GraphController,
+};
+use massa_models::{
+    api::BlockGraphStatus,
+    block::{BlockHeader, BlockId},
+    clique::Clique,
+    prehash::PreHashSet,
+    slot::Slot,
+    stats::ConsensusStats,
+    wrapped::Wrapped,
+};
+use massa_storage::Storage;
+use parking_lot::RwLock;
+use std::sync::{mpsc::SyncSender, Arc};
+
+use crate::{commands::GraphCommand, state::GraphState};
+
+#[derive(Clone)]
+pub struct GraphControllerImpl {
+    command_sender: SyncSender<GraphCommand>,
+    shared_state: Arc<RwLock<GraphState>>,
```

Comment:
I think you should avoid mixing channels and locks, because the two will not synchronize and it could easily lead to confusing behavior. 

For example, if you send a `GraphCommand::RegisterBlock(id)`, and then do `get_block_statuses(vec![id])`, even if it's from the same thread, you will sometimes still get `NotFound` back. 

The solution is to either do everything with channels(using oneshots within commands to get replies when needed), or do everything with a lock/condvar combo such as `Arc<(Condvar, RwLock<(GraphState, Option<GraphCommand>)>)>`

Reply:
This is done mostly to be able to:
* send commands through the channel without waiting for them to be processed from the point of view of the sending thread, and channels are very much optimal for that (much faster than locks)
* still be able to read the current state of the graph as processed so far (for this we need a shared state)

Note that sending commands and reading the state is done from different, mutually-asynchronous tasks.

But I agree that some documentation warning that the two are not in sync is important.